src/main/java/com/rocketden/main/service/ProblemService.java

```diff
@@ -83,6 +86,71 @@ public ProblemDto getProblem(String problemId) {
         return ProblemMapper.toDto(problem);
     }
 
+    public ProblemDto editProblem(String problemId, ProblemDto updatedProblem) {
+        Problem problem = repository.findProblemByProblemId(problemId);
+
+        if (problem == null) {
+            throw new ApiException(ProblemError.NOT_FOUND);
+        }
+
+        if (updatedProblem.getName() == null || updatedProblem.getDescription() == null
+                || updatedProblem.getDifficulty() == null
+                || updatedProblem.getProblemInputs() == null
+                || updatedProblem.getOutputType() == null) {
+            throw new ApiException(ProblemError.EMPTY_FIELD);
+        }
+
+        if (updatedProblem.getDifficulty() == ProblemDifficulty.RANDOM) {
+            throw new ApiException(ProblemError.BAD_DIFFICULTY);
+        }
+
+        // Ensure that the user entered valid inputs and outputs for the problem
+        for (ProblemTestCaseDto input : updatedProblem.getTestCases()) {
+            validateGsonParseable(input.getInput(), updatedProblem.getProblemInputs());
+            validateGsonParseable(input.getOutput(), updatedProblem.getOutputType());
+        }
+
+        problem.setName(updatedProblem.getName());
+        problem.setDescription(updatedProblem.getDescription());
+        problem.setDifficulty(updatedProblem.getDifficulty());
+        problem.setOutputType(updatedProblem.getOutputType());
+
+        problem.getProblemInputs().clear();
+        for (ProblemInputDto problemInput : updatedProblem.getProblemInputs()) {
+            if (problemInput != null) {
+                problem.addProblemInput(ProblemMapper.toProblemInputEntity(problemInput));
+            } else {
+                throw new ApiException(ProblemError.BAD_INPUT);
+            }
+        }
+
+        for (ProblemTestCaseDto testCaseDto : updatedProblem.getTestCases()) {
+            ProblemTestCase testCase = new ProblemTestCase();
+            testCase.setInput(testCaseDto.getInput());
+            testCase.setOutput(testCaseDto.getOutput());
+            testCase.setHidden(testCaseDto.isHidden());
+            testCase.setExplanation(testCaseDto.getExplanation());
+
+            problem.addTestCase(testCase);
```

Comment:
Should we add any checks here, or no need? For example, we could verify that this can be properly converted into the relevant types using the same method `validateGsonParseable`, while still just setting the problem with this the input string, output string, etc.

Reply:
Or, perhaps another solution is to create and save the problem, then call `createTestCase` with the relevant inputs and adding it through that method.

Reply:
I think it should already be verified above at line 109, although lmk if I missed any situation or I should move that block down or something (maybe that helps with readability?)

Reply:
Ah, I see -- makes sense thank you!